lib/chef/provider/template/content.rb

```diff
@@ -36,7 +36,26 @@ def template_location
         private
 
         def file_for_provider
-          context = TemplateContext.new(new_resource.variables)
+          # Deal with any DelayedEvaluator values in the template variables.
+          visitor = lambda do |obj|
+            case obj
+            when Hash
+              # Avoid mutating hashes in the resource in case we're changing anything.
+              obj.each_with_object(obj.class.new) do |(key, value), memo|
+                memo[key] = visitor.call(value)
+              end
+            when Array
+              # Avoid mutating arrays in the resource in case we're changing anything.
+              obj.map { |value| visitor.call(value) }
+            when DelayedEvaluator
+              new_resource.instance_eval(&obj)
+            else
+              obj
+            end
+          end
```

Comment:
+1 overall but will need to think through the API commonalities, not sure how much more readable a mixin-d version would end up looking vs. this. Could probably work out a few flavors though, value-only, dup style, and map! style?